rules/handler.go

```diff
@@ -0,0 +1,391 @@
+	errNoSnapshots         = errors.New("namespace has no snapshots")
+	errNamespaceTombstoned = errors.New("namespace is tombstoned")
+
+	namespaceActionErrorFmt = "cannot %s namespace %s. %v"
+
+	// Ruleset Errors
+	errNoUpdate      = errors.New("no new config given")
+	errNotTombstoned = errors.New("not tombstoned")
+
+	ruleActionErrorFmt    = "cannot %s rule %s. %v"
+	ruleSetActionErrorFmt = "cannot %s ruleset %s. %v"
+)
+
+// HandlerOptions holds common settings to all handlers.
+type HandlerOptions struct {
+	propagationDelay time.Duration
+}
+
+// NewHandlerOptions creates a new HanlderOptions struct.
+func NewHandlerOptions(propDelay time.Duration) HandlerOptions {
+	return HandlerOptions{propagationDelay: propDelay}
+}
+
+// MappingRuleConfig is the parts of a mapping rule that an end user would care about.
+// This struct is provided to create or make updates to the mapping rule.
+type MappingRuleConfig struct {
+	Name     string            `json:"name" validate:"nonzero"`
+	Filters  map[string]string `json:"filters" validate:"nonzero"`
+	Policies []policy.Policy   `json:"policies" validate:"nonzero"`
+}
+
+// MappingRuleUpdate contains a MappingRuleConfig along with an ID for a mapping rule.
+// The rule with that ID is meant to updated to the given config.
+type MappingRuleUpdate struct {
+	Config MappingRuleConfig `json:"config" validate:"nonzero"`
+	Name   string            `json:"name" validate:"nonzero"`
+}
+
+// RollupRuleConfig is the parts of a rollup rule that an end user would care about.
+// This struct is provided to create or make updates to the rollup rule.
+type RollupRuleConfig struct {
+	Name    string            `json:"name" validate:"nonzero"`
+	Filters map[string]string `json:"filters" validate:"nonzero"`
+	Targets []RollupTarget    `json:"targets" validate:"nonzero"`
+}
+
+// RollupRuleUpdate is a RollupRuleConfig along with an ID for a rollup rule.
+// The rule with that ID is meant to updated with a given config.
+type RollupRuleUpdate struct {
+	Config RollupRuleConfig `json:"config" validate:"nonzero"`
+	Name   string           `json:"name" validate:"nonzero"`
+}
+
+// RuleSetHandler exposes all of the methods to work with a RuleSet
+type RuleSetHandler interface {
+	// AppendMappingRule creates a new mapping rule and adds it to this ruleset.
+	AddMappingRule(MappingRuleConfig) error
+
+	// UpdateMappingRule creates a new mapping rule and adds it to this ruleset.
+	UpdateMappingRule(MappingRuleUpdate) error
+
+	// DeleteMappingRule deletes a mapping rule
+	DeleteMappingRule(ruleName string) error
+
+	// AppendRollupRule creates a new mapping rule and adds it to this ruleset.
+	AddRollupRule(RollupRuleConfig) error
+
+	// UpdateRollupRule creates a new mapping rule and adds it to this ruleset.
+	UpdateRollupRule(RollupRuleUpdate) error
+
+	// DeleteRollupRule deletes a rollup rule
+	DeleteRollupRule(ruleName string) error
+
+	// Tombstone tombstones this ruleset and all of its rules.
+	Delete() error
```

Comment:
Does this deletes the namespace too or is that handled in the namespace handler?